File: sdks/python/src/echo_agent_sdk/subagent_values.py
from __future__ import annotations

from enum import Enum


class SubagentCommandPhase(str, Enum):
    PERSISTED = "persisted"
    MAILBOX_ACCEPTED = "mailbox_accepted"
    DRAINED = "drained"
    TURN_SETTLED = "turn_settled"

    def as_str(self) -> str:
        return self.value

    @classmethod
    def parse(cls, value: str) -> SubagentCommandPhase | None:
        if not isinstance(value, str):
            return None
        try:
            return cls(value)
        except ValueError:
            return None


class SubagentStatus(str, Enum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    TIMED_OUT = "timed_out"

    def as_str(self) -> str:
        return self.value

    @classmethod
    def parse(cls, value: str) -> SubagentStatus:
        if not isinstance(value, str):
            raise TypeError("subagent status must be text")
        try:
            return cls(value)
        except ValueError as error:
            raise ValueError(f"unknown Subagent status: {value}") from error

File: sdks/typescript/src/subagent_values.ts
/** Subagent command/status values projected without owning dispatch state. */
export const SubagentCommandPhase = Object.freeze({
  Persisted: "persisted",
  MailboxAccepted: "mailbox_accepted",
  Drained: "drained",
  TurnSettled: "turn_settled",
} as const);

export type SubagentCommandPhase =
  (typeof SubagentCommandPhase)[keyof typeof SubagentCommandPhase];

export function subagentCommandPhaseAsStr(phase: SubagentCommandPhase): string {
  if (!Object.values(SubagentCommandPhase).includes(phase)) {
    throw new TypeError("invalid subagent command phase");
  }
  return phase;
}

export function subagentCommandPhaseParse(value: string): SubagentCommandPhase | undefined {
  if (typeof value !== "string") return undefined;
  return Object.values(SubagentCommandPhase).find((phase) => phase === value);
}

export const SubagentStatus = Object.freeze({
  Running: "running",
  Completed: "completed",
  Failed: "failed",
  Cancelled: "cancelled",
  TimedOut: "timed_out",
} as const);

export type SubagentStatus = (typeof SubagentStatus)[keyof typeof SubagentStatus];

export function subagentStatusAsStr(status: SubagentStatus): string {
  if (!Object.values(SubagentStatus).includes(status)) {
    throw new TypeError("invalid subagent status");
  }
  return status;
}

export function subagentStatusParse(value: string): SubagentStatus {
  if (typeof value !== "string") throw new TypeError("subagent status must be text");
  const status = Object.values(SubagentStatus).find((candidate) => candidate === value);
  if (status === undefined) throw new Error(`unknown Subagent status: ${value}`);
  return status;
}
